infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,150 @@
+package org.corfudb.infrastructure.logreplication;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationEvent;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationState;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.IllegalTransitionException;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeEvent;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeState;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeStateType;
+import org.corfudb.runtime.LogReplication.LogReplicationSession;
+
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * This class manages the tasks submitted by the runtime FSM and the replication FSM.
+ * Both the FSMs have their own instances of this class to which tasks are submitted and executed.
+ *
+ * The threads are shared by all the sessions traversing an FSM.
+ */
+@Slf4j
+public class FsmTaskManager {
+
+    private ExecutorService fsmWorker;
+
+    private Map<LogReplicationSession, LinkedList<UUID>> sessionToRuntimeEventIdMap;
+
+    private Map<LogReplicationSession, LinkedList<UUID>> sessionToReplicationEventIdMap;
```

Comment:
Shouldn't the value part be thread safe data structures as well? like some runtime events could come from `LogReplicationServer` that is now multithreaded, or at least we don't have control on the access path that could be in multithreaded environment?

Reply:
The handling of the linkedList was thread-safe except for the `add()` operation in addTask. Have changed that bit to use "computeIfPresent". This operation is blocking for ConcurrentHashMap